In [1]:
# Import libraries 
import pandas as pd 
import numpy as np
import os

In [3]:
# string
path =r'/Users/juliapetke/Desktop/Career Foundry/Instacart Basket Analysis'

In [92]:
# import of Products.csv
df_prods = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'products.csv'), index_col = False)

In [104]:
# import of orders_wrangled.csv
df_ords = pd.read_csv(os.path.join(path, '02 Data', 'Prepared Data', 'orders_2_wrangled.csv'), index_col = False)

In [106]:
df_ords.describe()

Unnamed: 0      order_id       user_id  order_number  \
count  3.421083e+06  3.421083e+06  3.421083e+06  3.421083e+06   
mean   1.710541e+06  1.710542e+06  1.029782e+05  1.715486e+01   
std    9.875817e+05  9.875817e+05  5.953372e+04  1.773316e+01   
min    0.000000e+00  1.000000e+00  1.000000e+00  1.000000e+00   
25%    8.552705e+05  8.552715e+05  5.139400e+04  5.000000e+00   
50%    1.710541e+06  1.710542e+06  1.026890e+05  1.100000e+01   
75%    2.565812e+06  2.565812e+06  1.543850e+05  2.300000e+01   
max    3.421082e+06  3.421083e+06  2.062090e+05  1.000000e+02   

       orders_day_of_week  order_time_of_day  days_since_prior_order  
count        3.421083e+06       3.421083e+06            3.214874e+06  
mean         2.776219e+00       1.345202e+01            1.111484e+01  
std          2.046829e+00       4.226088e+00            9.206737e+00  
min          0.000000e+00       0.000000e+00            0.000000e+00  
25%          1.000000e+00       1.000000e+01            4.000000e+00  
50%          3.000000e+00       1.300000e+01            7.000000e+00  
75%          5.000000e+00       1.600000e+01            1.500000e+01  
max          6.000000e+00       2.300000e+01            3.000000e+01

In [126]:
#overwriting a dataframe to create a new version with a different name in df_ords
df_ords = df_ords.drop(columns = ['Unnamed: 0'])

In [128]:
df_ords.describe()

order_id       user_id  order_number  orders_day_of_week  \
count  3.421083e+06  3.421083e+06  3.421083e+06        3.421083e+06   
mean   1.710542e+06  1.029782e+05  1.715486e+01        2.776219e+00   
std    9.875817e+05  5.953372e+04  1.773316e+01        2.046829e+00   
min    1.000000e+00  1.000000e+00  1.000000e+00        0.000000e+00   
25%    8.552715e+05  5.139400e+04  5.000000e+00        1.000000e+00   
50%    1.710542e+06  1.026890e+05  1.100000e+01        3.000000e+00   
75%    2.565812e+06  1.543850e+05  2.300000e+01        5.000000e+00   
max    3.421083e+06  2.062090e+05  1.000000e+02        6.000000e+00   

       order_time_of_day  days_since_prior_order  
count       3.421083e+06            3.214874e+06  
mean        1.345202e+01            1.111484e+01  
std         4.226088e+00            9.206737e+00  
min         0.000000e+00            0.000000e+00  
25%         1.000000e+01            4.000000e+00  
50%         1.300000e+01            7.000000e+00  
75%         1.600000e+01            1.500000e+01  
max         2.300000e+01            3.000000e+01

In [131]:
# 3.Check for mixed-type data in your df_ords dataframe.


In [164]:
df_ords_test = pd.DataFrame()

In [166]:
df_ords_test['mix'] = ['a', 'b', 1, True]

In [168]:
df_ords_test.head ()

mix
0     a
1     b
2     1
3  True

In [172]:
#checking for mix data type
for col in df_ords_test.columns.tolist():
    # Check if the column has mixed types by comparing the type of each element to the first element's type
    weird = (df_ords_test[col].map(type) != type(df_ords_test[col].iloc[0])).any()

#4. there are no mixed data types.


In [158]:
#5. Missing Values 

In [162]:
#finding missing values in df_ords
df_ords.isnull().sum()

order_id                       0
user_id                        0
order_number                   0
orders_day_of_week             0
order_time_of_day              0
days_since_prior_order    206209
dtype: int64

In [178]:
#creating a subset of the dataframe with the missing values
df_nan = df_ords[df_ords['days_since_prior_order'].isnull() == True]

In [180]:
df_nan 

order_id  user_id  order_number  orders_day_of_week  \
0         2539329        1             1                   2   
11        2168274        2             1                   2   
26        1374495        3             1                   1   
39        3343014        4             1                   6   
45        2717275        5             1                   3   
...           ...      ...           ...                 ...   
3420930    969311   206205             1                   4   
3420934   3189322   206206             1                   3   
3421002   2166133   206207             1                   6   
3421019   2227043   206208             1                   1   
3421069   3154581   206209             1                   3   

         order_time_of_day  days_since_prior_order  
0                        8                     NaN  
11                      11                     NaN  
26                      14                     NaN  
39                      11                     NaN  
45                      12                     NaN  
...                    ...                     ...  
3420930                 12                     NaN  
3420934                 18                     NaN  
3421002                 19                     NaN  
3421019                 15                     NaN  
3421069                 11                     NaN  

[206209 rows x 6 columns]

#The days_since_prior_order column contains missing values (NaN) because it represents the number of days since a user's previous order. For users placing their first order, there is no prior order to calculate this gap, leading to missing data. The dataset consists of 206,209 rows, with missing values occurring consistently in rows where the order number is 1, indicating these are users' first orders.


In [184]:
#6. Address the missing values using an appropriate method.

In [266]:
#Renaming Nan to O
df_nan = df_ords[df_ords['days_since_prior_order'].isnull() == False]

In [272]:
df_nan


order_id  user_id  order_number  orders_day_of_week  \
0         2539329        1             1                   2   
1         2398795        1             2                   3   
2          473747        1             3                   3   
3         2254736        1             4                   4   
4          431534        1             5                   4   
...           ...      ...           ...                 ...   
3421078   2266710   206209            10                   5   
3421079   1854736   206209            11                   4   
3421080    626363   206209            12                   1   
3421081   2977660   206209            13                   1   
3421082    272231   206209            14                   6   

         order_time_of_day  days_since_prior_order  
0                        8                     0.0  
1                        7                    15.0  
2                       12                    21.0  
3                        7                    29.0  
4                       15                    28.0  
...                    ...                     ...  
3421078                 18                    29.0  
3421079                 10                    30.0  
3421080                 12                    18.0  
3421081                 12                     7.0  
3421082                 14                    30.0  

[3421083 rows x 6 columns]

#Overall, mean imputation was chosen for its effectiveness in managing missing values without significantly distorting the dataset. While there are other methods available, such as median or mode imputation, mean imputation serves as a straightforward initial approach that can be refined later if necessary.

In [280]:
# 7. Run a check for duplicate values in your df_ords data.

In [292]:
#checking for duplicates in df_ords data
df_dups = df_ords_clean[df_ords_clean.duplicated()]

#there are no duplicates 

In [294]:
df_prods.to_csv(os.path.join(path, '02 Data','Prepared Data', 'prods_4.5.csv'))

In [296]:
df_ords.to_csv(os.path.join(path, '02 Data','Prepared Data', 'ords_4.5.csv'))